packages/flutter_tools/lib/src/cache.dart

```diff
@@ -1487,3 +1387,148 @@ const List<List<String>> _dartSdks = <List<String>> [
   <String>['linux-x64', 'dart-sdk-linux-x64.zip'],
   <String>['windows-x64', 'dart-sdk-windows-x64.zip'],
 ];
+
+/// An API for downloading and un-archiving artifacts, such as engine binaries or
+/// additional source code.
+class ArtifactUpdater {
+  ArtifactUpdater({
+    @required OperatingSystemUtils operatingSystemUtils,
+    @required Net net,
+    @required Logger logger,
+    @required FileSystem fileSystem,
+    @required Directory tempStorage
+  }) : _operatingSystemUtils = operatingSystemUtils,
+       _net = net,
+       _logger = logger,
+       _fileSystem = fileSystem,
+       _tempStorage = tempStorage;
+
+  final Logger _logger;
+  final Net _net;
+  final OperatingSystemUtils _operatingSystemUtils;
+  final FileSystem _fileSystem;
+  final Directory _tempStorage;
+
+  /// Keep track of the files we've downloaded for this execution so we
+  /// can delete them after completion. We don't delete them right after
+  /// extraction in case [update] is interrupted, so we can restart without
+  /// starting from scratch.
+  @visibleForTesting
+  final List<File> downloadedFiles = <File>[];
+
+  /// Download a zip archive from the given [url] and unzip it to [location].
+  Future<void> downloadZipArchive(
+    String message,
+    Uri url,
+    Directory location,
+  ) {
+    return _downloadArchive(
+      message,
+      url,
+      location,
+      _operatingSystemUtils.unzip,
+    );
+  }
+
+  /// Download a gzipped tarball from the given [url] and unpack it to [location].
+  Future<void> downloadZippedTarball(String message, Uri url, Directory location) {
+    return _downloadArchive(
+      message,
+      url,
+      location,
+      _operatingSystemUtils.unpack,
+    );
+  }
+
+  /// Download an archive from the given [url] and unzip it to [location].
+  Future<void> _downloadArchive(
+    String message,
+    Uri url,
+    Directory location,
+    void Function(File, Directory) extractor,
+  ) async {
+    final String downloadPath = flattenNameSubdirs(url, _fileSystem);
+    final File tempFile = _createDownloadFile(downloadPath);
+    final Status status = _logger.startProgress(
+      message,
+      timeout: null, // This will take a variable amount of time based on network connectivity.
+    );
+    int retries = 2;
+
+    while (retries > 0) {
+      try {
+        _ensureExists(tempFile.parent);
+        await _net.fetchUrl(url, destFile: tempFile, maxAttempts: 2);
+      } finally {
+        status.stop();
+      }
+      _ensureExists(location);
+
+      try {
+        extractor(tempFile, location);
+      } on ProcessException {
+        if (retries == 0) {
+          rethrow;
+        }
+        retries -= 1;
```

Comment:
The decrement should go before the `if (retries == 0)` I think.

Reply:
woops!